### PR TITLE
Feature/extend trajectory evaluation

### LINKED ISSUE #85: Expand trajectory evaluation and Streamlit review to evo artifact parity
# Minimal Benchmark-Ready Trajectory Evaluation Plan

## Summary

Build the first patch around cross-run benchmarking and clean trajectory contracts, not full evo plot artifact parity. The trajectory stage computes all required evo APE/RPE deviations, persists an aggregation-first manifest, and the rebuilt Streamlit Metrics page renders multi-run benchmark plots from persisted data only.

Benchmark rule: evaluate candidate trajectories against the selected GT/reference trajectory. ADVIO benchmarks vSLAM, ARKit, and ARCore against GT. TUM-RGBD benchmarks vSLAM against GT. Record3D benchmarks vSLAM against native ARKit-as-GT and does not emit an ARKit-vs-ARKit ranking row.

## Eval Stage And Contracts: `@ValentinBumeder`

- Replace the single `trajectory_metrics.json` contract with canonical trajectory evaluation artifacts:
  - `evaluation/trajectory/manifest.json`
  - `evaluation/trajectory/metrics_long.csv`
  - `evaluation/trajectory/error_series/*.npz`
  - optional short-lived legacy `evaluation/trajectory_metrics.json` alias only if needed for migration.
- Compute these evo-backed metrics for every candidate/reference pair:
  - APE: `full_transformation`, `translation_part`, `rotation_part`, `rotation_angle_rad`, `rotation_angle_deg`, `point_distance`.
  - RPE: all APE pose relations plus `point_distance_error_ratio`.
  - RPE default: `delta=1.0`, `delta_unit=frames`, `rel_delta_tol=0.1`, `all_pairs=false`, `pairs_from_reference=false`.
- Use evo native concepts as the source of truth:
  - `TrajectoryMetricFamily` maps to `evo.core.metrics.APE/RPE`.
  - `TrajectoryPoseRelation` mirrors `evo.core.metrics.PoseRelation`.
  - `RpeDeltaPolicy` mirrors evo RPE constructor fields.
  - `MetricStats` stays, but is constructed only from `metric.get_all_statistics()`.
- Persist row keys suitable for DataFrame indexing: `run_id`, `sequence_id`, `reference_source`, `estimate_source`, `metric_family`, `pose_relation`, `delta`, `delta_unit`, `statistic`.
- Persist the exact aligned/matched trajectory paths used for metrics. Do not report aligned stats while pointing the app back to raw estimates.
- Replace trajectory DTOs in `eval/contracts.py`:
  - Delete `TrajectorySeries`, `ErrorSeries`, `TrajectoryEvaluationPreview`, `TrajectoryEvaluationSemantics`, and `EvaluationArtifact`.
  - Replace them with `TrajectoryEvaluationManifest`, `TrajectoryMetricResultRow`, `TrajectoryMetricSpec`, `TrajectoryErrorSeriesRef`, and focused path/provenance DTOs.
  - Move trajectory alignment DTOs into a dedicated alignment contract module.
  - Leave cloud and intrinsics DTOs temporarily in `eval.contracts`; mark them as legacy cleanup follow-up, not part of trajectory eval.

## Service Boundaries And Run Orchestration

- Split `TrajectoryEvaluationService` responsibilities:
  - Stage/eval computation service: computes metrics from explicit prepared references and candidate trajectory refs.
  - App/post-run query service: discovers runs, loads manifests, builds aggregation tables, and never computes evo metrics.
- Change source-prepared benchmark inputs to separate roles:
  - `reference_trajectories`: GT/reference trajectories.
  - `candidate_trajectories`: external baseline trajectories to benchmark.
- Source preparation behavior:
  - ADVIO: GT in `reference_trajectories`; ARKit and ARCore in `candidate_trajectories`.
  - TUM-RGBD: GT in `reference_trajectories`; no external candidates by default.
  - Record3D: native ARKit in `reference_trajectories`; no ARKit self-candidate.
- Trajectory evaluation runtime adds normalized vSLAM `slam/trajectory.tum` as a candidate and evaluates every candidate against the selected reference.
- Stage outcome and summary manifests expose `trajectory_evaluation_manifest`, `metrics_long`, and error-series refs so downstream app pages aggregate without recomputation.

## Streamlit Metrics Page

- Rebuild `src/prml_vslam/app/pages/metrics.py` as the benchmark aggregation page.
- Selection flow: dataset -> sequence -> all matching run roots/methods.
- The page reads persisted manifests and `metrics_long.csv`; it must not call evo compute paths.
- Default views:
  - ranked metrics table by method/source/metric/stat;
  - APE position-error ranking emphasized for benchmark readability;
  - absolute position-error CDFs matching ADVIO reporting style;
  - box and histogram plots from persisted error-series refs;
  - filters for metric family, pose relation, reference source, estimate source, and run.
- Remove the pipeline-side APE preview path and its eval DTO dependency. Pipeline may link to persisted trajectory evaluation artifacts after the stage completes.

## Tests

- Synthetic evo metric tests for APE and 1-frame RPE across translation, rotation angle, and full transformation where evo supports it.
- Manifest round-trip tests for `manifest.json`, `metrics_long.csv`, and `.npz` error-series refs.
- Source/orchestration tests:
  - ADVIO produces GT reference plus ARKit/ARCore candidates.
  - TUM-RGBD produces GT reference only.
  - Record3D produces ARKit-as-GT reference only and no self-ranking row.
- Runtime tests proving stage artifacts include manifest, metrics table, and error-series refs.
- Streamlit model/page tests proving multi-run aggregation reads persisted artifacts and does not invoke evo computation.
- Contract cutover tests proving trajectory consumers no longer import deleted DTOs from `eval.contracts`.

## Assumptions

- Stage-generated static PNG/HTML evo plots are out of this patch; Streamlit owns interactive plotting from persisted metric data.
- Cloud and intrinsics contract cleanup is deferred.
- The first implementation may keep a legacy `trajectory_metrics.json` compatibility alias, but the canonical contract is the trajectory manifest plus long-form metrics table.

### PR DESCRIPTION
**Pull request by ValentinBumeder**
**feat: dataset-wide trajectory metric aggregation**

Extends the metrics page from single-sequence review to a full dataset-wide benchmark dashboard.

**Evaluation (pipeline)**
Trajectory evaluation now computes 4 metrics per candidate: APE translation, APE rotation, RPE translation (1 m delta), RPE rotation (1 m delta). Non-primary failures are silently skipped.
Error-series paths in metrics_long.csv are now stored relative (error_series/filename.npz), making artifacts portable across machines. Legacy absolute paths from other machines are remapped on load.

**New: dataset aggregation layer (eval/dataset_aggregation.py)**
Pure functions (no I/O): build_per_sequence_table, build_leaderboard, build_coverage_matrix, build_heatmap_data
Types: MetricFilter, PerSequenceRow, LeaderboardRow, CoverageMatrix, HeatmapData

**New: dataset query helpers (eval/query.py)**
DatasetEvaluationSelection, DatasetRunCoverage
discover_dataset_runs, load_dataset_evaluation, load_dataset_coverage

**New: plot builders (plotting/metrics.py)**
build_dataset_heatmap, build_grouped_bar_per_sequence, build_coverage_chart, build_violin_by_method

**Metrics page (app/pages/metrics.py)**
New scope toggle: Single Sequence (existing behaviour) / Dataset Overview
Dataset view: coverage heatmap, leaderboard table, per-sequence heatmap, grouped bar, violin

Tests: 39 new tests across test_dataset_aggregation, test_dataset_query, test_dataset_plots; test_eval updated for multi-metric assertions. Full suite: 531 passed.

### CODE REVIEW

File: src/prml_vslam/eval/services.py [hunk 1]
```diff
@@ -377,102 +271,122 @@ def compute_pipeline_evaluation(
...
+                            reference_path=reference_path,
+                            estimate_path=candidate.path,
+                            pose_relation=spec.pose_relation,
+                            delta=spec.delta or 1.0,
+                            delta_unit=spec.delta_unit_enum or metrics.Unit.meters,
+                        )
+                except ValueError:
+                    if spec.family == "ape" and spec.pose_relation is metrics.PoseRelation.translation_part:
+                        raise
+                    continue
```
Comment (JanDuchscherer104) [inline src/prml_vslam/eval/services.py:334]: This `continue` silently drops every non-primary metric failure. If APE rotation or either RPE metric fails, the manifest and `metrics_long.csv` give no indication that the metric was attempted and failed, so the app can present the run as complete with missing metrics. Please persist these failures in the trajectory manifest and surface a failure count/message through the query layer and metrics page status UI, while still raising for the primary APE translation case.

File: src/prml_vslam/eval/dataset_aggregation.py [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+
+
+def build_heatmap_data(
+    rows: list[PerSequenceRow],
+    all_sequence_ids: list[str],
+    metric_name: str = "APE RMSE (m)",
+) -> HeatmapData:
+    """Build a heatmap matrix of values indexed by sequence × estimate source."""
+    estimate_sources = sorted({f"{r.estimate_source_base}/{r.coordinate_status}" for r in rows})
+    value_map: dict[tuple[str, str], float] = {
```
Comment (JanDuchscherer104) [inline src/prml_vslam/eval/dataset_aggregation.py]: This key collapses all runs for the same `(sequence_id, estimate_source_base/coordinate_status)`. Re-running the same method on the same sequence overwrites the earlier value in the heatmap, and the leaderboard later treats duplicate rows as separate sequence contributions. The per-sequence row needs to retain a stable run identity, and aggregation/display labels need to be duplicate-aware instead of keying only on source.

File: src/prml_vslam/plotting/metrics.py [hunk 3]
```diff
@@ -49,3 +67,162 @@ def build_error_figure(error_series: ErrorSeries) -> go.Figure:
...
+    )
+    return figure
+
+
+def build_grouped_bar_per_sequence(rows: list[PerSequenceRow]) -> go.Figure:
+    """Build a grouped bar chart of metric values by sequence and estimate source."""
+    groups: dict[str, dict[str, float]] = defaultdict(dict)
+    for row in rows:
+        source_label = f"{row.estimate_source_base}/{row.coordinate_status}"
+        groups[source_label][row.sequence_id] = row.value
```
Comment (JanDuchscherer104) [inline src/prml_vslam/plotting/metrics.py]: The grouped bar chart has the same collision: `groups[source_label][row.sequence_id] = row.value` overwrites duplicate runs for a sequence/source pair. This makes repeated runs disappear from the dataset view. Please use the same duplicate-aware source/run label strategy here as in the heatmap and cover it with a regression test.

**Review by JanDuchscherer104 — CHANGES_REQUESTED** [review]
Blocking review from the PR #95 deep pass. Please do not merge until these are resolved:

1. Non-primary trajectory metric failures are currently dropped during manifest generation. That makes rotation/RPE failures invisible in the persisted artifacts and in the Streamlit metrics page, so a run can look complete while important metrics never computed.
2. Dataset aggregation/chart labels are keyed only by sequence plus estimate source. Duplicate runs for the same sequence/source overwrite each other in heatmaps and grouped bars, and leaderboard counts become run-count based instead of sequence-aware.
3. Rerun evaluation logging still needs distinct metric identities and scalar leaves for APE translation, APE rotation, and RPE cases. Copilot already left inline comments on `src/prml_vslam/visualization/rerun_policy.py` around `_evaluation_metric_id()` and `/error/translation_m`; those are merge-blocking too.

I validated these issues in a separate worktree and have local fixes/tests for all three areas. Minimum acceptance before merge: persist metric failures in the manifest/query/app status path, carry a stable run identity through metric rows and dataset aggregation, make duplicate display labels collision-safe, and add focused regression tests for Rerun metric identity plus duplicate-run aggregation/chart behavior.

File: src/prml_vslam/plotting/metrics.py (hunk 3 above)
Comment (ValentinBumeder) [reply src/prml_vslam/plotting/metrics.py]: Fixed in commit "fix: fixed duplicate run bug in plotting/metrics.py"

File: src/prml_vslam/eval/trajectory_contracts.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+        if isinstance(value, str) and value in metrics.PoseRelation.__members__:
+            return metrics.PoseRelation[value]
+        return value
+
+    @field_serializer("pose_relation", when_used="json")
+    def _serialize_pose_relation(self, value: metrics.PoseRelation) -> str:
+        return value.name
+
+    """Number of associated pose pairs used by the metric."""
+
```
[the remark this answers is not in the document]
Comment (ValentinBumeder) [reply src/prml_vslam/eval/trajectory_contracts.py]: Fixed with commit "fix: fixed docstring in eval/services.py"

File: src/prml_vslam/eval/services.py (hunk 1 above)
Comment (ValentinBumeder) [reply src/prml_vslam/eval/services.py:334]: Fixed with commit "fix: fixed continue in eval/services.py dropping metric failures."

File: src/prml_vslam/eval/dataset_aggregation.py (hunk 2 above)
Comment (ValentinBumeder) [reply src/prml_vslam/eval/dataset_aggregation.py]: Fixed with commit "fix: fixed duplicate run bug in dataset_aggregation.py"
